implemented quantize_per_tensor_dynamic and added a corresponding test script

dzdang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #68004

Differential Revision: [D32301792](https://our.internmc.facebook.com/intern/diff/D32301792)

Commit:
implemented quantize_per_tensor_dynamic and added a corresponding test script

[ghstack-poisoned]

jerryzh168 commented on test/quantization/core/test_quantized_tensor.py:
```diff
@@ -143,9 +143,27 @@ class TestQuantizedTensor(TestCase):
+            scale, zero_pt = _calculate_dynamic_qparams(mat2quant, dtype, reduce_flag)
+            q_s = torch.quantize_per_tensor(mat2quant, scale, zero_pt, dtype)
+
+            print("Original tensor: \n", mat2quant)
```
we can use self.assertEqual(q_d, q_s) to check they match I think

dzdang replied:
I'll add that in. I wasn't sure how the backend floating point arithmetic worked with the quantization and whether the two will actually end up equal or not equal for all use cases due to precision issues

dzdang commented on aten/src/ATen/native/quantized/QTensor.cpp:
```diff
@@ -11,6 +11,42 @@
+  int qmax = 255;
+
+  // should we check additional QInt types beyond 8 bits?
+  if (dtype == ScalarType::QInt8) {
+    qmin = -128;
+    qmax = 127;
+  } else if (dtype == ScalarType::QUInt8) {
+  } else {
+    // dtype not supported
+  }
```
I was considering doing the same as https://github.com/pytorch/pytorch/blob/master/test/quantization/core/test_quantized_tensor.py#L31, but in https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/quant_utils.h#L70, the `reduce_flag` is processed, so I think since I'm calling quant_utils::ChooseQuantizationParams, then I don't need to manually process the reduce_flag?

jerryzh168 replied:
oh I see, yeah that sounds good, please put qmin/qmax assignment for QUInt8 to the else branch though for better readability.

Also we can add a `TORCH_ASSERT` for the dtype: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/conv_serialization.h#L284

Commit:
Update on "implemented quantize_per_tensor_dynamic and added a corresponding test script"




[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/quantized/QTensor.cpp:
```diff
@@ -11,6 +11,43 @@
+  } else if (dtype == ScalarType::QUInt8) {
+    qmin = 0;
+    qmax = 255;
+  }
```
nit: we can add a comment here saying that we should add new branches when we add support for other dtypes

dzdang replied:
Done

jerryzh168 commented on aten/src/ATen/native/quantized/QTensor.cpp:
```diff
@@ -11,6 +11,43 @@
+    reduce_range = false;
+  }
+
+  int qmin = 0;
+  int qmax = 255;
```
nit: maybe remove the initializations since qmin and qmax are determined by dtype

dzdang replied:
I'll make the change. For future reference, is the practice at FB / PyTorch to leave variables uninitialized if you know that it'll be set later (before it's used)? I'm used to initializing everything :)

jerryzh168 replied:
I don't remember if there is a practice like that, I feel this should be OK. There are also cases when it does make sense to initialize like: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/q_avgpool.cpp#L406, so I think leave things uninitialized can happen sometimes.

For this particular case, since the choice of the initialized value is a bit arbitrary I feel it is clearer just to leave it uninitialized.

jerryzh168 commented on test/quantization/core/test_quantized_tensor.py:
```diff
@@ -143,9 +143,33 @@ class TestQuantizedTensor(TestCase):
+            scale, zero_pt = _calculate_dynamic_qparams(mat2quant, dtype, reduce_flag)
+            q_s = torch.quantize_per_tensor(mat2quant, scale, zero_pt, dtype)
+
+            print("Original tensor: \n", mat2quant)
+            print("Dynamically quantized tensor: \n", q_d)
+            print("Statically quantized tensor: \n", q_s)
+            print("Scale = ", scale, "Zero point = ", zero_pt)
```
can we remove the prints?

dzdang replied:
Will do

jerryzh168 reviewed (approved):
Looks good, thanks!

Commit:
Update on "implemented quantize_per_tensor_dynamic and added a corresponding test script"




[ghstack-poisoned]

Commit:
Update on "implemented quantize_per_tensor_dynamic and added a corresponding test script"




[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32301792).

jerryzh168 commented on aten/src/ATen/native/quantized/QTensor.cpp:
```diff
@@ -11,6 +11,44 @@
+  int qmin;
+  int qmax;
+
+  // should we check additional QInt types beyond 8 bits?
```
nit: we can remove this

Commit:
Update on "implemented quantize_per_tensor_dynamic and added a corresponding test script"


Differential Revision: [D32301792](https://our.internmc.facebook.com/intern/diff/D32301792)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32301792).

Commit:
Update on "implemented quantize_per_tensor_dynamic and added a corresponding test script"


Differential Revision: [D32301792](https://our.internmc.facebook.com/intern/diff/D32301792)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32301792).

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32301792).

Commit:
Update on "implemented quantize_per_tensor_dynamic and added a corresponding test script"


Differential Revision: [D32301792](https://our.internmc.facebook.com/intern/diff/D32301792)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32301792).

Commit:
Update on "implemented quantize_per_tensor_dynamic and added a corresponding test script"


Differential Revision: [D32301792](https://our.internmc.facebook.com/intern/diff/D32301792)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32301792).

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32301792).

bdhirsh commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -5310,6 +5310,11 @@
+- func: quantize_per_tensor_dynamic(Tensor self, ScalarType dtype, bool reduce_range) -> Tensor
+  variants: function
+  dispatch:
+    CPU, CUDA: quantize_per_tensor_dynamic
```
Hey @jerryzh168! This PR adds `torch.quantize_per_tensor_dynamic` to the python frontend, but it doesn't have any associated API docs. For the 1.11 release, we want to either remove the op from python, or add some public docs.

If you want to add docs for it, can you make a patch and cherry-pick it into the 1.11 release branch?

Otherwise if you're okay with keeping it out of python, let me know and I can make a patch.

jerryzh168 replied:
sure, I think we can add a doc for this, when is the deadline for cherrypick?

bdhirsh replied:
@jerryzh168 AFAIK there isn't a deadline yet, but you can track the progress here (or ask questions): https://fb.workplace.com/groups/671272153461300/permalink/999082684013577/
